src/HDF5.jl

```diff
@@ -1484,21 +1484,31 @@ do_reclaim(::Type{T}) where T <: Union{Cstring, VariableArray} = true
 
 function read(dset::HDF5Dataset, T::Union{Type{Array{U}}, Type{U}}) where U <: NamedTuple
   filetype = datatype(dset)
-  memtype_id = h5t_get_native_type(filetype.id)  # padded layout in memory
-  @assert sizeof(U) == h5t_get_size(memtype_id) "Type sizes mismatch!"
+  memtype = HDF5Datatype(h5t_get_native_type(filetype.id))  # padded layout in memory
+  close(filetype)
+
+  if sizeof(U) != h5t_get_size(memtype.id)
+    h5type_str = h5lt_dtype_to_text(memtype.id)
+    error("""
```

Comment:
When I try to trigger this error (without the corresponding fix to reading nested types)

I get an error on the julia size from `sizeof(T)`: 
`Argument is an incomplete Array type and does not have a definite size.`

Reply:
Good catch. Should have been `U` instead of `T`. Should be fixed now.